### PR TITLE
Remove backend dependencies from the connection code

### PR DESCRIPTION
**Pull request by flatcap**
<details>
<summary>Object Diagram</summary>

</details>

Each backend Account holds a Connection, which in turn holds a ConnAccount.
- `struct Connection` manages the socket handling
- `struct ConnAccount` is a cache of login credentials

The cached credentials allow NeoMutt to poll all the Accounts without having to fire `account-hook`s every time.

For each credential (User, Password, etc), there is some backend-specific code that looks up which config variable to use.
- `$imap_user`, `$pop_user`, ...
- `$imap_pass`, `$pop_pass`, ...

---

This brings SMTP into line with IMAP and POP.

- cd8bb0e6c config: add missing `$smtp_user` variable

Minor tidying to make the code easier to read:

- 2a439d873 rename: unify ConnAccount variables
- d4d295b84 initialise all ConnAccount variables
- dee5b7a0b drop unused param in mutt_conn_find()

Ages ago, I put a `ConnAccount` into the POP and IMAP Mailbox data.  I can't remember why.
This _may_ have been causing connection problems as some code updated the wrong credentials cache.

- 14c1fb343 imap: drop ConnAccount
- db200311a pop: drop ConnAccount

Add a callback function to the `ConnAccount`, so that it can ask the backends for credentials.

- c1db14f87 refactor connection service
- c5f18d437 refactor: add get_field() to Connection

Finally, move some dependency-free functions to the connection library.

- 7acfbcecf conn: move ConnAccount functions to libconn
  - `mutt_account_getlogin()`
  - `mutt_account_getoauthbearer()`
  - `mutt_account_getpass()`
  - `mutt_account_getuser()`
  - `mutt_account_unsetpass()`

### CODE REVIEW

**Comment by Austin-Ray** [comment]
I'll make sure to review once you've fixed what's going on. Just ping me when you're ready.

**Comment by flatcap** [comment]
@neomutt/reviewers 
OK, this is ready to be reviewed.

**Comment by gahr** [comment]
Will try to review within a couple of days!

File: pop/pop_lib.c [hunk 1]
```diff
@@ -51,45 +51,70 @@
...
+  switch (field)
+  {
+    case MUTT_CA_USER:
+      return C_PopUser;
+    case MUTT_CA_PASS:
+      return C_PopPass;
+    case MUTT_CA_OAUTH_CMD:
+      return C_PopOauthRefreshCommand;
+    case MUTT_CA_HOST:
+    case MUTT_CA_LOGIN:
```
Comment (gahr) [inline pop/pop_lib.c]: Should this return C_PopUser, too?

File: conn/connaccount.c [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+  if (!token || (*token == '\0'))
+  {
+    mutt_error(_("Command returned empty string"));
+    FREE(&token);
+    return NULL;
+  }
+
+  /* Determine the length of the keyed message digest, add 50 for overhead. */
+  size_t oalen = strlen(cac->login) + strlen(cac->host) + strlen(token) + 50;
```
[the remark this answers is not in the document]
Comment (gahr) [reply conn/connaccount.c]: Oh, but we have `mutt_str_asprintf` too :)

File: conn/connaccount.c [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  return 0;
+}
+
+/**
+ * mutt_account_unsetpass - Unset ConnAccount's password
+ * @param cac ConnAccount to modify
+ */
+void mutt_account_unsetpass(struct ConnAccount *cac)
+{
+  cac->flags &= ~MUTT_ACCT_PASS;
```
[the remark this answers is not in the document]
Comment (flatcap) [reply conn/connaccount.c]: Done.
I'd also like to turn the `char[]` into `char*` at some point (but not today),
which would mean we could get rid of the flags.

File: conn/connaccount.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (flatcap) [reply conn/connaccount.c]: Nice.  Done.

File: imap/imap.c [hunk 4]
```diff
@@ -1859,13 +1859,12 @@ static struct Account *imap_ac_find(struct Account *a, const char *path)
...
-  struct ConnAccount *ac = &adata->conn_account;
+  struct ConnAccount *cac = &adata->conn->account;
 
-  if ((mutt_str_strcasecmp(url->host, ac->host) != 0) ||
-      (mutt_str_strcasecmp(url->user, ac->user) != 0))
-  {
+  if (mutt_str_strcasecmp(url->host, cac->host) != 0)
+    a = NULL;
+  else if (url->user && (mutt_str_strcasecmp(url->user, cac->user) != 0))
     a = NULL;
```
[the remark this answers is not in the document]
Comment (flatcap) [reply imap/imap.c]: The original code expects both **host** and **user** to be present on both sides.
This isn't always the case.
The new check is more relaxed: if the canonical path has a user, but the mailbox path doesn't, that's OK.

Hopefully, my upcoming **path** changes will help sort this out -- we'll always have canonical paths to work with.

File: conn/connaccount.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (flatcap) [reply conn/connaccount.c]: Hmm... now LGTM isn't happy

File: pop/pop_lib.c (hunk 1 above)
Comment (flatcap) [reply pop/pop_lib.c]: Hmm... good spot.  This is a bit more interesting.
My first draft of the code had `MUTT_CA_LOGIN` return `C_PopUser`.
The first time I tested the Pop changes, it hung for ~5 seconds, then succeeded.

The 'login' field is only used by SASL.
For me, sending 'flatcap' causes a hang, sending `NULL` works instantly.

I'll investigate further...

Comment (flatcap) [reply pop/pop_lib.c]: Argh!  `ETOOMUCHCONFIG`
Login now returns `C_PopUser` as it should.

**Pull request merged**